### PR TITLE
Quantity enhancements and cash prototype

### LINKED ISSUE #9: Money System
**Summarize your proposal**
A system to store the amount of money a player has

**Describe the solution you'd like**
A central system should hold the amount of money any player has at any given time
The system needs to provide multiple interfaces for server-driven and player-driven transactions
The Amount of money needs to be stored persistently 
If money should exist as real items or just a virtual number needs to be defined.

### PR DESCRIPTION
**Pull request by Arkensor**
- Add cash prototype
- Add inventory utils that can handle quantity prefabs too
- Add script component finder by other component utility
- Add reusable number abbreviation format util
- Update quantity system to expose change invoker and tidy up API
- Update inventory utils to handle null to become easier to chain when using
- Update default loadout to include $100
- Update HUD to display cash

Cash has no volume and no weight to be put into any ever-so-tiny storage, and "full" inventories. Otherwise, large amounts of money would take up the majority of the character's inventory. 
image

Closes #9

### CODE REVIEW

File: src/Scripts/Game/Feature/Quantity/EL_InventoryUtils.c [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	//------------------------------------------------------------------------------------------------
+	//! \ref GetAmount(InventoryStorageManagerComponent, ResourceName)
+	static int GetAmount(IEntity target, ResourceName prefab)
+	{
+		InventoryStorageManagerComponent invManager = EL_InventoryUtils.GetResponsibleStorageManager(target);
+		return GetAmount(invManager, prefab);
+	}
+
+	//------------------------------------------------------------------------------------------------
+	//! Returns acutal amount added or 0 on faulty opertations.
```
Comment (thebonbon) [inline src/Scripts/Game/Feature/Quantity/EL_InventoryUtils.c:36]:
```suggestion
	//! Returns actual amount added or 0 on faulty operations.
```

File: src/Scripts/Game/Feature/Quantity/EL_InventoryUtils.c [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	static int AddAmount(InventoryStorageManagerComponent storageManager, ResourceName prefab, int amount, bool dropOverflow = false)
+	{
+		if (!storageManager || amount <= 0) return 0;
+
+		int remainainAmount = amount;
```
Comment (thebonbon) [inline src/Scripts/Game/Feature/Quantity/EL_InventoryUtils.c:41]:
```suggestion
		int remainingAmount = amount;
```

File: src/Scripts/Game/Feature/Quantity/EL_InventoryUtils.c [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		return AddAmount(invManager, prefab, amount, dropOverflow);
+	}
+
+	//------------------------------------------------------------------------------------------------
+	//! Returns acutal amount removed or 0 on faulty opertations.
+	static int RemoveAmount(InventoryStorageManagerComponent storageManager, ResourceName prefab, int amount)
+	{
+		if (!storageManager || amount <= 0) return 0;
+
+		int remainainAmount = amount;
```
Comment (thebonbon) [inline src/Scripts/Game/Feature/Quantity/EL_InventoryUtils.c:98]:
```suggestion
		int remainingAmount = amount;
```

File: src/Scripts/Game/Feature/Quantity/ScriptedInventoryStorageManagerComponent.c [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+			int newTotal = EL_InventoryUtils.GetAmount(this, prefab);
+			if (newTotal == -1) continue;
+
+			int previous = m_mELAmountCache.Get(prefab);
+			m_mELAmountCache.Set(prefab, newTotal);
+
+			ScriptInvoker invoker = m_mELItemCountInvokers.Get(prefab);
+			if (invoker && newTotal != previous) invoker.Invoke(prefab, newTotal, previous);
+
+			PrintFormat(ToString()+"::OnAmountChanged(%1, %2, %3)", prefab, newTotal, previous);
```
Comment (thebonbon) [inline src/Scripts/Game/Feature/Quantity/ScriptedInventoryStorageManagerComponent.c:49]: Is this debug print needed?

**Review by thebonbon — CHANGES_REQUESTED** [review]

File: src/Scripts/Game/HUD/EL_BetaHud.c [hunk 5]
```diff
@@ -121,16 +121,16 @@ class EL_BetaHud : SCR_InfoDisplay
...
+		if (!m_MoneyIndicator)
 		{
 			m_MoneyIndicator = TextWidget.Cast(m_PlayerStatsHUD.FindAnyWidget("m_moneyIndicator"));
 			if (!m_MoneyIndicator) return;
 		}
-		
-		m_MoneyIndicator.SetText("$ " + value); //for configurabiluity could have the $ changeable in config...
+
+		m_MoneyIndicator.SetText("$ " + EL_FormatUtils.AbbreviateNumber(newAmount)); //for configurabiluity could have the $ changeable in config...
 		//m_StatChange = true;
```
Comment (thebonbon) [inline src/Scripts/Game/HUD/EL_BetaHud.c:134]: > //m_StatChange = true;

I dont know how this works but after a change in money the UI should be visible again

Comment (Arkensor) [reply src/Scripts/Game/HUD/EL_BetaHud.c:134]: Yes, probably, but the hud will need to be redone. I just put money in there temporarily.

**Pull request merged**